<img style="float: left;padding: 1.3em" src="https://indico.in2p3.fr/event/18313/logo-786578160.png">  

#  Gravitational Wave Open Data Workshop #3


#### Tutorial 1.1: Discovering open data from GW observatories

This notebook describes how to discover what data are available from the [Gravitational-Wave Open Science Center (GWOSC)](https://www.gw-openscience.org).
    
[Click this link to view this tutorial in Google Colaboratory](https://colab.research.google.com/github/gw-odw/odw-2020/blob/master/Day_1/Tuto%201.1%20Discovering%20Open%20Data.ipynb)

## Software installation  (execute only if running on a cloud platform or haven't done the installation yet!)

First, we need to install the software, which we do following the instruction in [Software Setup Instructions](https://github.com/gw-odw/odw-2020/blob/master/setup.md):

In [1]:
# -- Uncomment following line if running in Google Colab
!pip install -U gwosc

**Important:** With Google Colab, you may need to restart the runtime after running the cell above.

In [2]:
#check the version of the package gwosc you are using
import gwosc
print(gwosc.__version__)

0.8.1


The version you get should be 0.8 or greater. If it's not, check that you have followed all the steps in [Software Setup Instructions](https://github.com/gw-odw/odw-2020/blob/master/setup.md).

## Querying for event information

The module `gwosc.catalog`  provides tools to search for events in a catalog.

The module `gwosc.datasets` provides tools for searching for datasets, including full run strain data releases.

For example, we can search for events in the [GWTC-1 catalog](https://www.gw-openscience.org/eventapi/html/GWTC-1-confident/), the catalog of all events from the O1 and O2 observing runs.  A list of available catalogs can be seen in the [Event Portal](https://gw-openscience.org/eventapi)

In [3]:
from gwosc.datasets import find_datasets

# -- GW events del catálogo GWTC-1-confident
gwtc1 = find_datasets(type="event", catalog="GWTC-1-confident")
print("GWTC-1 events:", gwtc1)
print("")

# -- Conjuntos grandes de datos de strain (runs de LIGO/Virgo)
runs = find_datasets(type="run")
print("Large data sets:", runs)

GWTC-1 events: ['GW150914-v3', 'GW151012-v3', 'GW151226-v2', 'GW170104-v2', 'GW170608-v3', 'GW170729-v1', 'GW170809-v1', 'GW170814-v3', 'GW170817-v3', 'GW170818-v1', 'GW170823-v1']

Large data sets: ['O1', 'O2', 'O3GK', 'O3a', 'O3b', 'O4a', 'O4b1Disc', 'O4b2Disc', 'O4b3Disc', 'S5', 'S6']


`catalog.events` also accepts a `segment` and `detector` keyword to narrow results based on GPS time and detector:

In [4]:
#-- Detector and segments keywords limit search results

datasets = find_datasets(
    type="event",
    catalog="GWTC-1-confident",
    detector="L1",
    segment=(1164556817, 1187733618)
)

print(datasets)

['GW170104-v2', 'GW170608-v3', 'GW170729-v1', 'GW170809-v1', 'GW170814-v3', 'GW170817-v3', 'GW170818-v1', 'GW170823-v1']


Using `gwosc.datasets.event_gps`, we can query for the GPS time of a specific event:

In [5]:
from gwosc.datasets import event_gps
gps = event_gps('GW190425')
print(gps)

1240215503.0


<div class="alert alert-info">All of these times are returned in the GPS time system, which counts the number of seconds that have elapsed since the start of the GPS epoch at midnight (00:00) on January 6th 1980. GWOSC provides a <a href="https://www.gw-openscience.org/gps/">GPS time converter</a> you can use to translate into datetime, or you can use <a href="https://gwpy.github.io/docs/stable/time/"><code>gwpy.time</code></a>.</div>

We can query for the GPS time interval for an observing run:

In [7]:
from gwosc.datasets import run_segment
print(run_segment('O1'))

(1126051217, 1137254417)


To see only the confident events in O1:

In [8]:
# Segmento GPS del run O1 (2015–2016)
O1_segment = (1126051217, 1137254417)

# Buscar los eventos del catálogo GWTC-1-confident que ocurrieron en O1
O1_events = find_datasets(
    type="event",
    catalog="GWTC-1-confident",
    segment=O1_segment
)

print("Eventos en O1:", O1_events)


Eventos en O1: ['GW150914-v3', 'GW151012-v3', 'GW151226-v2']


## Querying for data files

The `gwosc.locate` module provides a function to find the URLs of data files associated with a given dataset.

For event datasets, one can get the list of URLs using only the event name:

In [9]:
from gwosc.locate import get_event_urls
urls = get_event_urls('GW150914')
print(urls)

['https://gwosc.org/archive/data/O1/1126170624/H-H1_LOSC_4_V1-1126256640-4096.hdf5', 'https://gwosc.org/archive/data/O1/1126170624/L-L1_LOSC_4_V1-1126256640-4096.hdf5']


By default, this function returns all of the files associated with a given event, which isn't particularly helpful. However, we can can filter on any of these by using keyword arguments, for example to get the URL for the 32-second file for the LIGO-Livingston detector:

In [10]:
urls = get_event_urls('GW150914', duration=32, detector='L1')
print(urls)

['https://gwosc.org/archive/data/O1/1126170624/L-L1_LOSC_4_V1-1126256640-4096.hdf5']


# Exercises

Ahora que has visto ejemplos de cómo consultar información sobre conjuntos de datos usando el paquete `gwosc`, intenta completar los siguientes ejercicios utilizando esa interfaz:

* ¿Cuántos meses duró S6?

* ¿Cuántos eventos del catálogo GWTC-1-confident fueron detectados durante O1?

* ¿Qué URL de archivo contiene los datos para el detector V1 en un intervalo de 4096 segundos alrededor de GW170817?

# SOLUCIÓN

## ¿Cuántos meses duró S6?

In [13]:
from gwosc.datasets import run_segment
print(run_segment('S6'))

(931035615, 971622015)


Al realizar la conversión del tiempo GPS al tiempo del calendario gregoriano se tiene:

* 931035615 = 2009-07-07 T 21:00:00
* 971622015 = 2010-10-20 T 15:00:00

Por lo tanto S6 duró 1 año y 3 meses

## ¿Cuántos eventos del catálogo GWTC-1-confident fueron detectados durante O1?

In [14]:
from gwosc.datasets import run_segment
print(run_segment('O1'))

(1126051217, 1137254417)


In [15]:
# Segmento GPS del run O1 (2015–2016)
O1_segment = (1126051217, 1137254417)

# Buscar los eventos del catálogo GWTC-1-confident que ocurrieron en O1
O1_events = find_datasets(
    type="event",
    catalog="GWTC-1-confident",
    segment=O1_segment
)

print("Eventos en O1:", O1_events)

Eventos en O1: ['GW150914-v3', 'GW151012-v3', 'GW151226-v2']


Solo hay 3 eventos registrados para ese segmento de tiempo

## ¿Qué URL de archivo contiene los datos para el detector V1 en un intervalo de 4096 segundos alrededor de GW170817?

In [16]:
from gwosc.locate import get_event_urls
urls = get_event_urls('GW170817')
print(urls)

['https://gwosc.org/eventapi/json/GWTC-1-confident/GW170817/v3/H-H1_GWOSC_4KHZ_R1-1187008867-32.hdf5', 'https://gwosc.org/eventapi/json/GWTC-1-confident/GW170817/v3/H-H1_GWOSC_4KHZ_R1-1187006835-4096.hdf5', 'https://gwosc.org/eventapi/json/GWTC-1-confident/GW170817/v3/L-L1_GWOSC_4KHZ_R1-1187008867-32.hdf5', 'https://gwosc.org/eventapi/json/GWTC-1-confident/GW170817/v3/L-L1_GWOSC_4KHZ_R1-1187006835-4096.hdf5', 'https://gwosc.org/eventapi/json/GWTC-1-confident/GW170817/v3/V-V1_GWOSC_4KHZ_R1-1187008867-32.hdf5', 'https://gwosc.org/eventapi/json/GWTC-1-confident/GW170817/v3/V-V1_GWOSC_4KHZ_R1-1187006835-4096.hdf5', 'https://gwosc.org/eventapi/json/GWTC-1-confident/GW170817/v3/G-G1_GWOSC_4KHZ_R1-1187008867-32.hdf5', 'https://gwosc.org/eventapi/json/GWTC-1-confident/GW170817/v3/G-G1_GWOSC_4KHZ_R1-1187006835-4096.hdf5']


In [17]:
urls = get_event_urls('GW170817', duration=4096, detector='V1')
print(urls)

['https://gwosc.org/eventapi/json/GWTC-1-confident/GW170817/v3/V-V1_GWOSC_4KHZ_R1-1187006835-4096.hdf5']


Al realizar el filtro de la duración de los 4096 segundos y el detector V1 para el evento GW170817 tenemos que el url del archivo que contiene los datos es:

['https://gwosc.org/eventapi/json/GWTC-1-confident/GW170817/v3/V-V1_GWOSC_4KHZ_R1-1187006835-4096.hdf5']

# Respuestas a preguntas planteadas